pkg/reconciler/buildrun/resources/params.go

```diff
@@ -53,3 +65,511 @@ func overrideParams(originalParams []buildv1alpha1.ParamValue, overrideParams []
+			// not documented use case is that a Build specifies a value for an optional parameter (which has a default in the strategy). Then one can
+			// set a paramValue without a value in the BuildRun to get back to the default value.
+			return nil
+		}
+
+	case buildv1alpha1.ParameterTypeArray:
+		taskRunParam.Value.Type = pipeline.ParamTypeArray
+
+		if paramValue.Values != nil {
+			for index, value := range paramValue.Values {
+				if value.ConfigMapValue != nil {
+					envVarName, err := addConfigMapEnvVar(taskRun, paramValue.Name, value.ConfigMapValue.Name, value.ConfigMapValue.Key)
+					if err != nil {
+						return err
+					}
+
+					envVarExpression := fmt.Sprintf("$(%s)", envVarName)
+					if value.ConfigMapValue.Format != nil {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, strings.ReplaceAll(*value.ConfigMapValue.Format, "${CONFIGMAP_VALUE}", envVarExpression))
+					} else {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, envVarExpression)
+					}
+				} else if value.SecretValue != nil {
+					envVarName, err := addSecretEnvVar(taskRun, paramValue.Name, value.SecretValue.Name, value.SecretValue.Key)
+					if err != nil {
+						return err
+					}
+
+					envVarExpression := fmt.Sprintf("$(%s)", envVarName)
+					if value.SecretValue.Format != nil {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, strings.ReplaceAll(*value.SecretValue.Format, "${SECRET_VALUE}", envVarExpression))
+					} else {
+						taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, envVarExpression)
+					}
+				} else if value.Value != nil {
+					taskRunParam.Value.ArrayVal = append(taskRunParam.Value.ArrayVal, *value.Value)
+				} else {
+					// this error should never happen because we validate this upfront in ValidateBuildRunParameters
+					return fmt.Errorf("unexpected parameter without any value: %s[%d]", parameterDefinition.Name, index)
+				}
+			}
+		} else if parameterDefinition.Defaults == nil {
+			// this error should never happen because we validate this upfront in ValidateBuildRunParameters
+			return fmt.Errorf("unexpected parameter without any value: %s", parameterDefinition.Name)
+		} else {
+			// we tolerate this for optional parameters, we enter this code path if a user provides a paramValue without any values. The theoretic but
+			// not documented use case is that a Build specifies values for an optional parameter (which has defaults in the strategy). Then one can
+			// set a paramValue without values in the BuildRun to get back to the default values.
+			return nil
+		}
+	}
+
+	taskRun.Spec.Params = append(taskRun.Spec.Params, taskRunParam)
+
+	return nil
+}
+
+func generateEnvVarName(prefix string) (string, error) {
+	result := prefix
+
+	for i := 0; i < 5; i++ {
+		num, err := rand.Int(rand.Reader, EnvVarNameSuffixCharsCount)
+		if err != nil {
+			return "", err
+		}
+		result += string(EnvVarNameSuffixChars[num.Int64()])
```

Comment:
interesting approach ... borrowed from somewhere, or an original from you?

I suspect it in part stems from the restrictions / constratins on env var names.  Perhaps a comment detailing the highlights of that would be good.

WDYT?

Reply:
What exactly surprised you?

I defined in the [ship that the environment variable gets a random suffix](https://github.com/shipwright-io/community/blob/main/ships/0025-strategy-parameters-enhancements.md#backend). And because tools like gosec complain every random number usage that is not using `crypto/rand`, I used `crypto/rand`. Unfortunately that beast only has a signature to work with `big.Int`s rather than primitive ints, which makes the code look a little more complex than needed, especially at the top when `EnvVarNameSuffixCharsCount` is initialized. If you're aware of a simpler solution, let me know.

I admin that I actually did not check which restrictions on environment variable names exist. I just used uppercase letters and digits for simplicity.

BTW: the purpose of the random suffix is not to obfuscate something, it is rather that I need an environment variable name that is unique.

Reply:
yeah "surprise" was the wrong choice of words ... its connotation was not what I was wanting to convey.

let me try again

the algorithm was novel and new to me;  that is not at all a bad thing; in fact, there is novelty and inventiveness in it that I'm trying to convey; I was more curious, because of the novelty, if you derived it entirely by yourself, or if you were inspired from somewhere else in the k8s ecosystem, that is all

hopefully addressing that piece, even if the env var name restrictions did not come into play for you, I think it is an aspect which your algorithm achieves, and could be part of a comment you add.  re: restrictions, see https://github.com/kubernetes/api/blob/107b3100d94652e198a0e866ed0c2d028e6649f0/core/v1/types.go#L197 and https://github.com/kubernetes/kubernetes/blob/f2ddd60eb9e7e9e29f7a105a9a8fa020042e8e52/pkg/api/v1/types.go#L42-L4

some of your history of what was attempted previously could also be useful to new users who view the code in the future without knowing immediately to look for say the SHIP or something

anyway, if you are amenable to adding some comments, great; I won't block the PR if you feel strongly the code is self explanatory.